---
title: "R Notebook"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

```{r}
pacman::p_load(ape,
               ggtree,
               ggnewscale,
               phangorn,
               TreeTools,
               tidyverse,
               janitor,
               viridis)
```




```{r}
metadata <-
    read_csv("./docs/isolate_metadata.csv") %>%
    select(!c(genotype))


mykrobe <-
    read.table("./docs/mykrobe.out_predictResults.tsv",
    sep="\t", header=TRUE) %>%
    remove_empty() %>%
    remove_constant() %>%
    rename("accession_number" = "genome")

#merge the metadata file and mykrobe results
metadata <-
  base::merge(metadata, mykrobe, by = "accession_number",
        all.x = TRUE)

# group isolates into this study and others
metadata$group <-
  case_when(metadata$publication == "Jillian et al. 2021" ~ "This study",
  TRUE ~ "Other")

# group isolates locale into Malawi and other countries
metadata$origin_country <-
  case_when(metadata$country == "Malawi" ~ "Malawi",
  TRUE ~ "Other")


#write the metadata to disk
write_csv(metadata,
  file = "./docs/metadata_plus_mykrobe.csv")
```


```{r}

tree <-
  ape::read.tree("/Users/zuza/repos/H58/outputs/phylo.aln.treefile")

tree <- ladderize(tree, right = TRUE)

tree <- midpoint(tree)

tree <- reorder.phylo(tree, order = "cladewise")


#get tip labels in the current tree
labels <- TipLabels(tree)

#subset metadata file using the labels in the tree
data <- metadata %>% filter(accession_number %in% c(labels))



p <- ggtree(tree, layout = "rectangular",
  ladderize = TRUE)
p

# get the node ids for annotation later
nodeid(tree, label = c("ERR2602821", "ERR360832", "ERR279139", "ERR2602832", "ERR279153"))
```



### color the tips of the tree by country of sample isolation
```{r}
# color the tips of the tree
p1 <-
  p %<+% data +
  theme(legend.position = "none",
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        plot.title = element_text(
          size = 12,
          face = "bold",
          hjust = 0.5,
          vjust = -15
        )
  ) +
  geom_tippoint(aes(color=genotype),
                size = 0.5)

p1

```



### create a heatmap for the country

```{r}

# create heatmap with multiple variables
colors <- c("#f9f6f3", "#3a0177")
y_offset <- 3
outdir <- "outputs/plots"

#select columns with plasmids
plasmid <-
   data %>%
  select("accession_number", contains(c("IncFIAHI1", "IncHI1A", "IncHI1BR27")))


#set row names to the column for accession_number
rownames(plasmid) <- plasmid$accession_number

# remove accession number column so that the country variable
#has the accession variable as the row nakes
plasmid <- plasmid %>%
    select(!accession_number) %>%
    mutate_at(c("IncFIAHI1", "IncHI1A", "IncHI1BR27"),
    as.factor)


plasmid_plot <-
    gheatmap(p, plasmid,
        width = 0.08,
        colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = y_offset,
        hjust = 0.2,
        legend_title = "Plasmid Replicons") +
        scale_fill_manual(values = colors,
        aesthetics = "fill") +
    labs(fill = "Presence/Absence")

plasmid_plot

h1 <- plasmid_plot + new_scale_fill()


#select columns with amr genes
amr <-
   data %>%
  select("accession_number",
         "blaTEM.1D",
         "blaCTX.M.15",
         "catA1",
         "sul1",
         "sul2",
         "dfrA7",
         "dfrA14",
         "tetA",
         "tetB")
  
#set row names to the column for accession_number
rownames(amr) <- amr$accession_number

amr <- amr %>%
    select(!accession_number) 

genes <- colnames(amr)
amr <- amr %>% mutate_at(genes,
    as.factor)


amr_plot <-
    gheatmap(h1, amr,
        width = 0.36,
        offset = 0.0000001,
        colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = y_offset,
        hjust = 0.2,
        legend_title = "AMR genes") +
        scale_fill_manual(values = colors,
        aesthetics = "fill") +
        labs(fill = NULL) + theme(plot.margin = unit(c(3, 0.5, 0.5, 0.5), "cm"))

amr_plot

h2 <- amr_plot + new_scale_fill()

# qunolone sceptibility
mutations <-
   data %>%
  select("accession_number", contains(c("gyrA_S83F", "gyrB_S464F")))

rownames(mutations) <- mutations$accession_number

mutations <- mutations %>%
    select(!accession_number) 

points <- colnames(mutations)
mutations <- mutations %>% mutate_at(points,
    as.factor)


mutations_plot <-
    gheatmap(h2, mutations,
        width = 0.054,
        offset = 0.00000047,
        colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = y_offset,
        hjust = 0.2,
        legend_title = "AMR genes") +
        scale_fill_manual(values = colors,
        aesthetics = "fill") +
        labs(fill = NULL) + theme(plot.margin = unit(c(3, 0.5, 0.5, 0.5), "cm"))

mutations_plot
```


```{r}
# save the phages
plotting_dir <- file.path(outdir, Sys.Date())
plot_name <- file.path(plotting_dir, "jillian_tree_plot.svg")

if (!dir.exists(plotting_dir)) {dir.create(plotting_dir, recursive = TRUE)}

svg(plot_name,
    width = 9,
    height = 20)

mutations_plot + theme(plot.margin = unit(c(3, 0.5, 0.5, 0.5), "cm"))


dev.off() 



```
