---
title: "R Notebook"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

```{r}
# library(pak)
# pak("ggtree")
# pak("igraph")

library(ape)
library(ggtree)
library(ggnewscale)
library(phangorn)
library(tidyverse)
library(janitor)
library(plotly)
library(viridis)
library(tidytree)
library(ggbreak)
library(devtools)

# load the metadata
load(file = "./docs/metadata_plus_mykrobe.RData")


```



# process metadata
```{r combine_metadata_files, eval=FALSE}
metadata <-
    read_csv("./docs/isolate_metadata.csv") %>%
    select(!c(accession_number))

mykrobe <-
    read.table("./docs/mykrobe.out_predictResults.tsv",
    sep="\t", header=TRUE) %>%
    remove_constant() %>%
    rename("sample_name" = "genome")


ariba <-
  read_csv("./docs/ariba_clean.csv") %>%
  remove_constant() %>%
  rename("sample_name" = "sample_id")

#merge the metadata file and mykrobe results
metadata <-
  base::merge(metadata, mykrobe, by = "sample_name",
        all.x = TRUE)

metadata <-
  base::merge(metadata, ariba, by = "sample_name",
        all.x = TRUE)

# group isolates into this study and others
metadata$group <-
  case_when(metadata$publication == "Jillian et al. 2021" ~ "This study",
  TRUE ~ "Other")

# group isolates locale into Malawi and other countries
metadata$origin_country <-
  case_when(metadata$country == "Malawi" ~ "Malawi",
  TRUE ~ "Other")

# save metadata to object
save(metadata, file = "./docs/metadata_plus_mykrobe.RData")
save(ariba, file = "./docs/amr_genes.RData")

#write the metadata to disk
write_csv(metadata,
  file = "./docs/metadata_plus_mykrobe.csv")
```


# load newick files
```{r read_newick_files, eval=FALSE}
#read in the trees
big_tree <-
  ape::read.tree("~/Downloads/iq2.contree")

first_tree <-
  ape::read.tree("./outputs/newick/2024-04-19/phylo.aln.treefile")

malawi <-
  ape::read.tree("./outputs/iqtree/malawi/filtered_alignment.fas.treefile")

malawi_no_ref <-
  ape::read.tree("./outputs/iqtree/malawi/phylo.aln.treefile")

global <-
  ape::read.tree("./outputs/iqtree/global/filtered_alignment.fas.treefile")

global_no_ref <-
  ape::read.tree("./outputs/iqtree/global/phylo.aln.treefile")

# save to an R object
save(tree, malawi, malawi_no_ref, global, global_no_ref,
  file = "./data/r/trees.Rdata")
```

# load data
```{r}

load("./data/r/trees.Rdata")
load("./docs/metadata_plus_mykrobe.RData")
load("./docs/amr_genes.RData")

```



# color the tips of the tree by country of sample isolation
```{r}
# color the tips of the tree
data <- metadata
tree <- ape::read.tree("/Users/zuza/repos/H58/outputs/iqtree/20240629/iqtree/phylo.aln.contree")

tree %>% drop.tip("AL513382.1") %>% ggtree() %<+% data + aes(color = year)# + scale_colour_grey()

```

# annotation files
```{r}
# create heatmap with multiple variables
colors <- c("#f9f6f3", "#3a0177")

outdir <- "outputs/plots"


data <- metadata #%>% filter(sample_name %in% c(labels))

point_mutations <- 
  data %>% select("sample_name", matches(c("par", "gyr"))) %>% remove_constant()

rownames(point_mutations) <- point_mutations$sample_name

point_mutations <- point_mutations %>% select(!sample_name)

point_mutations <- point_mutations %>%
  mutate(across(where(is.integer), as.factor))

# create heatmap for country, and group
group <-
  data %>%
  select(sample_name, group) %>%
  mutate(group = factor(group))

rownames(group) <-
  group$sample_name

group <-
  group %>% select(!sample_name)

# create a country object
data$region <-
  case_when(data$country == "Malawi" ~ "Malawi",
            TRUE ~ "Other")

region <-
  data %>%
  select(sample_name, region) %>%
  mutate(region = factor(region))

rownames(region) <-
  region$sample_name

region <-
  region %>% select(!sample_name)


# heatmap for the year
year <-
  data %>%
  select(sample_name, year) %>%
  mutate(year = factor(year))


rownames(year) <- year$sample_name

year <- year %>% select(!sample_name)

# plasmids

#select columns with plasmids
plasmid_replicons <- 
  c("IncFIAHI1", "IncHI1A", "IncHI1BR27", "IncHI1_ST6", "IncY", "IncFIB_pHCM2",
    "IncFIB_K", "IncN")

plasmid <-
   data %>%
    select("sample_name", all_of(plasmid_replicons))

#set row names to the column for accession_number
rownames(plasmid) <- plasmid$sample_name

# remove accession number column so that the country variable
#has the accession variable as the row nakes
plasmid <- plasmid %>%
    select(!sample_name) %>%
    mutate_at(c("IncFIAHI1", "IncHI1A", "IncHI1BR27", "IncHI1_ST6", "IncY", "IncFIB_pHCM2",
    "IncFIB_K", "IncN"),
    as.factor)


#select columns with amr genes

genes <- colnames(ariba)
amr <-
   data %>% select(all_of(genes))
  
#set row names to the column for accession_number
rownames(amr) <- amr$sample_name

amr <- amr %>%
    mutate_at(genes,
    as.factor) %>%
    select(!sample_name)

amr[] <- lapply(amr, function(x) str_replace_all(x, "yes", "1"))
amr[] <- lapply(amr, function(x) str_replace_all(x, "no", "0"))

# merge the amr determinants for tree vis
amr_determinants <- 
  cbind(amr, point_mutations, plasmid)

```



# global isolates
this section will plot a tree for the global isolates.

```{r}
tree <- global
data <- metadata
#tree <- midpoint(tree)
tree <- root(tree, "AL513382.1")
tree <- reorder.phylo(tree, order = "cladewise")
#get tip labels in the current tree
#labels <- TipLabels(tree)

#write.csv(labels, file="./docs/tree_labels.csv")

#subset metadata file using the labels in the tree
#data <- metadata
# plot parameters
y_offset <- 15
vertical_expansion = c( 0.0002)
font_size <- 0.5


# # convert tree to tible for easy assessing
# tree %>% as.tibble() %>% filter(label %in% c("AL513382.1", "ERR279139"))
# 
# tree %>% as.tibble() %>% filter(node == "100/100")
# child(tree, 1204)
# 
# tree %>% ggtree() + geom_nodelab(nudge_y = -10, size = 1.4, nudge_x = -0.00004) #+ theme(element_text(size = 1))
# nodeid(tree, "AL513382.1")
# 
# pdf("test.pdf", width = 10,  height = 60)
# #p + geom_nodelab( size = 1.4)
# ggtree(y, aes(linetype = group)) + theme(legend.position = "none")# + scale_x_break(c(0.000034, 0.0000004)) #+ geom_tiplab(as_ylab = TRUE)
# dev.off()
# child(tree, "99.1/100")
# 
# MRCA(tree, 548, 547)
# 
# MRCA(tree, "99.1/100")
# y <- groupClade(tree, 1058)
# 
# x <- ggtree(y, aes(linetype = group), cex = .2) +scale_x_break(c(0.000034, 0.0000004)) #%<+% data
# 
# x
# 
# y <- ggtree(y, aes(linetype = group)) + theme(legend.position = "none") + scale_x_break(c(0.000034, 0.0000004)) %<+% data +
#   geom_tippoint(aes(color=genotype), size = 0.5)
# 
# ggtree(y, aes(linetype = group), cex = .2)
# y
# 
# p + scale_x_break()


p <- 
  tree %>%
  #drop.tip("AL513382.1") %>%
  ggtree(cex = .2) %<+% data +
  geom_tippoint(aes(color=genotype), size = 0.02) +
  scale_color_viridis_d(na.translate = FALSE,
                        guide = guide_legend(order = 1))

p

p3 <-
    gheatmap(p, year,
        width = 0.005 ,
        colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = y_offset,
        hjust = 0.5,
        font.size = font_size,
        legend_title = "Year") +
      scale_fill_manual(values = date_col,
                        guide = guide_legend(order = 2)) +
      labs(fill = "Year") +
      hexpand(.2) +
      vexpand(vertical_expansion) #+
      guides(fill=guide_legend(ncol = 1, byrow = TRUE)) 

p3

# add the country of isolation
p4 <- p3 + new_scale_fill()

p5 <-
    gheatmap(p4, group,
        width = 0.005 ,
        offset = 0.000001,
        colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = y_offset,
        hjust = 0.5,
        font.size = font_size,
        legend_title = "Group") +
  scale_fill_manual(values = group_col,
                    na.translate = F,
                    labels = c("Yes", "No"),
                    guide = guide_legend(order = 3)) +
  labs(fill = "This Study") +
  hexpand(.2) +
  vexpand(vertical_expansion)

p5

# add region column

p6 <- p5 + new_scale_fill()

p7 <-
    gheatmap(p6, region,
        width = 0.005 ,
        offset = 0.000002,
        colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = y_offset,
        hjust = 0.5,
        font.size = font_size,
        legend_title = "Region") +
  scale_fill_manual(values = region_col, na.translate = F,
                    guide = guide_legend(order = 4)) + 
   labs(fill = "Region") +
  hexpand(.2) +
  vexpand(vertical_expansion)
p7


p8 <- p7 + new_scale_fill()

p9 <-
    gheatmap(p8, amr_determinants,
        width = 0.2,
        offset = 0.000003,
        colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = y_offset,
        hjust = 0.5,
        font.size = font_size,
        legend_title = "AMR genes and Plasmids") +
  scale_fill_manual(values = amr_gene_col, na.translate = F,
                    labels = c("Absent", "Present"),
                    guide = guide_legend(order = 5)) +
    labs(fill = "AMR/Plasmid") +
  hexpand(.2) +
  vexpand(vertical_expansion) #+
  guides(fill=guide_legend(ncol = 1, byrow = TRUE))

p9 + theme(legend.key.size = unit(.1, "cm")) +
  theme_light(base_size = 5) +
  xlim(NA, 0.000155) +
  ylim(-0.000001, NA)


```


```{r}
# save the phages
plotting_dir <- file.path(outdir, Sys.Date())
plot_name <- file.path(plotting_dir, "global_tree.pdf")

if (!dir.exists(plotting_dir)) {dir.create(plotting_dir, recursive = TRUE)}

pdf(plot_name,
    width = 9,
    height = 20)

p9 + theme(legend.key.size = unit(.1, "cm")) +
  theme_light(base_size = 5) +
  xlim(NA, 0.000155) +
  ylim(-0.000001, NA)


dev.off() 


```


# Malawi only isolates

```{r}
tree <- malawi

tree <- root(tree, "AL513382.1")
tree <- reorder.phylo(tree, order = "cladewise")

#get tip labels in the current tree
labels <- TipLabels(tree)

malawi_amr <-
  subset(amr_determinants, rownames(amr_determinants) %in%
           tree$tip.label) %>% remove_constant()

# plot parameters
y_offset <- 8
vertical_expansion = c( 0.03)
font_size <- 2


# tree %>% as_tibble() %>% filter(parent == 315)#filter(label %in% c("AL513382.1", "ERR2602632"))
# 
# MRCA(tree, "88.6/98")
 
# tree
y <- groupClade(tree, "86.6/98")


# get nodeid based on tip label
nodes <- nodeid(tree, c("ERR2602821", "ERR2602832"))

p <- 
  ggtree(y) %<+% data + scale_x_break(c(0.00001, 0.00013), expand = F) +  hexpand(0.00009, direction = -1) + hexpand(0.09, direction = 1) +
  geom_highlight(node = nodes, fill = "#3a0177", alpha = .6, extend = 1)

p
p1 <-
    gheatmap(p, year,
        width = 0.007 ,
        offset = 0.00001,#0.000015,
        colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = y_offset,
        hjust = 0.5,
        font.size = font_size) +
      scale_fill_manual(values = c("2015" = "gray", "2016" = "black"),
                        guide = guide_legend(order = 2)) +
      labs(fill = "Year") +
      hexpand(0.2, direction = 1) +
      vexpand(vertical_expansion)
p1

# add the country of isolation
p2 <- p1 + new_scale_fill()

p3 <-
    gheatmap(p2, malawi_amr,
        width = 0.1 ,
        offset = 0.000012,
        colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = y_offset,
        #hjust = 0.5,
        font.size = font_size) +
  scale_fill_manual(values = amr_gene_col,
                    na.translate = F,
                    labels = c("Absent", "Present"),
                    guide = guide_legend(order = 3)) +
  labs(fill = "AMR genes and Plasmids") +
  vexpand(vertical_expansion) + xlim(-.0000005, 0.000145)

p3


p3 + theme(legend.key.size = unit(.1, "cm"))

```


```{r}
# save the phages
plotting_dir <- file.path(outdir, Sys.Date())
plot_name <- file.path(plotting_dir, "malawi_tree.pdf")

if (!dir.exists(plotting_dir)) {dir.create(plotting_dir, recursive = TRUE)}

pdf(plot_name,
    width = 10,
    height = 20)

p3 + theme(legend.key.size = unit(.1, "cm"))

dev.off() 

```


# Malawi only isolates

```{r}
tree <- first_tree

tree <- reorder.phylo(tree, order = "cladewise")

#get tip labels in the current tree
malawi_amr <-
  subset(amr_determinants, rownames(amr_determinants) %in%
           tree$tip.label) %>% remove_constant()

# plot parameters
y_offset <- 8
vertical_expansion = c( 0.03)
font_size <- 2


# tree %>% as_tibble() %>% filter(parent == 315)#filter(label %in% c("AL513382.1", "ERR2602632"))
# 
# MRCA(tree, "88.6/98")
 
# tree
y <- groupClade(tree, "86.6/98")


# get nodeid based on tip label
nodes <- nodeid(tree, c("ERR2602821", "ERR2602832"))

p <- 
  ggtree(tree) %<+% data +
  #scale_x_break(c(0.00001, 0.00013), expand = F) +  hexpand(0.00009, direction = -1) + hexpand(0.09, direction = 1) +
  geom_highlight(node = nodes, fill = "#3a0177", alpha = .6, extend = 1)

p
p1 <-
    gheatmap(p, year,
        width = 0.014 ,
        #offset = 0.00001,#0.000015,
        colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = y_offset,
        hjust = 0.5,
        font.size = font_size) +
      scale_fill_manual(values = c("2015" = "gray", "2016" = "black"),
                        guide = guide_legend(order = 2)) +
      labs(fill = "Year") +
      #hexpand(0.2, direction = 1) +
      vexpand(vertical_expansion)
p1

# add the country of isolation
p2 <- p1 + new_scale_fill()

p3 <-
    gheatmap(p2, malawi_amr,
        width = 0.2 ,
        offset = 0.000000021,
        colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = y_offset,
        #hjust = 0.5,
        font.size = font_size) +
  scale_fill_manual(values = amr_gene_col,
                    na.translate = F,
                    labels = c("Absent", "Present"),
                    guide = guide_legend(order = 3)) +
  labs(fill = "AMR genes and Plasmids") +
  vexpand(vertical_expansion) #+ xlim(-.0000005, 0.000145)

p3


p3 + theme(legend.key.size = unit(.1, "cm"))

```


```{r}
# save the phages
plotting_dir <- file.path(outdir, Sys.Date())
plot_name <- file.path(plotting_dir, "first_malawi_tree.pdf")

if (!dir.exists(plotting_dir)) {dir.create(plotting_dir, recursive = TRUE)}

pdf(plot_name,
    width = 10,
    height = 20)

p3 + theme(legend.key.size = unit(.1, "cm"))

dev.off() 

```


# extra stuff
```{r}
tree <- first_tree
tree <- midpoint(tree)
tree <- reorder.phylo(tree, order = "cladewise")

#get tip labels in the current tree
labels <- TipLabels(tree)

#write.csv(labels, file="./docs/tree_labels.csv")

#subset metadata file using the labels in the tree
data <- metadata %>% filter(sample_name %in% c(labels))

# plot parameters
y_offset <- 3
vertical_expansion = c( 0.0002)

# get nodeid based on tip label
nodes <- nodeid(tree, c("ERR2602821", "ERR2602832"))

p <-
  tree %>%
  drop.tip("AL513382.1") %>%
  ggtree(cex = 1) +
  geom_tiplab(align = TRUE,
              family="mono",
              size = 2,
              linesize = 0.2,
              as_ylab = FALSE) +
  geom_highlight(node = nodes,, fill = "#3a0177", alpha = .6, extend = 1)

p
p3 <-
    gheatmap(p, year,
        width = 0.02 ,
        offset = 0.0000004,
        colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = y_offset,
        hjust = 0.5,
        font.size = font_size,
        legend_title = "Plasmid Replicons") +
  scale_fill_manual(values = date_col) +
  labs(fill = "Year") +
  hexpand(.2) +
  vexpand(vertical_expansion) +
  theme(legend.justification = "left") +
  guides(fill=guide_legend(ncol = 1, byrow = TRUE))

p3

# extend the tree to include plasmid replicons and amr genes
mbe <-
  cbind(year,plasmid, amr) %>%
  select(c("IncFIAHI1", "IncHI1A", "IncHI1BR27","blaTEM.1D",
         "catA1",
         "sul1",
         "sul2",
         "dfrA7",
         "dfrA14",
         "tetB"))

p8 <- p3 + new_scale_fill()

p9 <-
    gheatmap(p8, mbe,
        width = 0.2,
        offset = 0.0000006,
        colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = y_offset,
        hjust = 0.5,
        font.size = font_size,
        legend_title = "Plasmid Replicons") +
  scale_fill_viridis_d(
        aesthetics = "fill",
        direction = -1) +
    labs(fill = "Presence/Absence") +
  hexpand(.2) +
  vexpand(vertical_expansion)

p9


# include point mutation
p10 <- p9 + new_scale_fill()

p11 <-
    gheatmap(p10, point_mutations %>% select(c("gyrA_S83F", "gyrB_S464F")),
        width = 0.07,
        offset = 0.0000014,
        colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = y_offset,
        hjust = 0.5,
        font.size = font_size,
        legend_title = "Plasmid Replicons") +
  scale_fill_viridis_d(aesthetics = "fill", direction = -1) +
    labs(fill = "Presence/Absence") +
  hexpand(.2) +
  vexpand(vertical_expansion)
p11

p11 + theme(legend.key.size = unit(1, "cm")) +
  theme_light(base_size = 10) +
  xlim(NA, 0.000006) +
  ylim(-0.000001, NA)

ggplotly(p3)
```




# malawi first tree
```{r}
tree <- first_tree
tree <- midpoint(tree)
tree <- reorder.phylo(tree, order = "cladewise")

#get tip labels in the current tree
labels <- TipLabels(tree)

#write.csv(labels, file="./docs/tree_labels.csv")

#subset metadata file using the labels in the tree
data <- metadata %>% filter(sample_name %in% c(labels))

# plot parameters
y_offset <- 3
vertical_expansion = c( 0.0002)

# get nodeid based on tip label
nodes <- nodeid(tree, c("ERR2602821", "ERR2602832"))

p <-
  tree %>%
  drop.tip("AL513382.1") %>%
  ggtree(cex = 1) +
  geom_tiplab(align = TRUE,
              family="mono",
              size = 2,
              linesize = 0.2,
              as_ylab = FALSE) +
  geom_highlight(node = nodes,, fill = "#3a0177", alpha = .6, extend = 1)

p
p3 <-
    gheatmap(p, year,
        width = 0.02 ,
        offset = 0.0000001,
        colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = y_offset,
        hjust = 0.5,
        font.size = font_size,
        legend_title = "Plasmid Replicons") +
  scale_fill_viridis_d(aesthetics = "fill") +
    labs(fill = "Year") +
  hexpand(.2) +
  vexpand(vertical_expansion) +
  theme(legend.justification = "left") + guides(fill=guide_legend(ncol = 1, byrow = TRUE))

p3

# extend the tree to include plasmid replicons and amr genes
mbe <-
  cbind(year,plasmid, amr) %>%
  select(c("IncFIAHI1", "IncHI1A", "IncHI1BR27","blaTEM.1D",
         "catA1",
         "sul1",
         "sul2",
         "dfrA7",
         "dfrA14",
         "tetB"))

p8 <- p3 + new_scale_fill()

p9 <-
    gheatmap(p8, mbe,
        width = 0.2,
        offset = 0.00000015,
        colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = y_offset,
        hjust = 0.5,
        font.size = font_size,
        legend_title = "Plasmid Replicons") +
  scale_fill_viridis_d(
        aesthetics = "fill",
        direction = -1) +
    labs(fill = "Presence/Absence") +
  hexpand(.2) +
  vexpand(vertical_expansion)

p9


# include point mutation
p10 <- p9 + new_scale_fill()

p11 <-
    gheatmap(p10, point_mutations %>% select(c("gyrA_S83F", "gyrB_S464F")),
        width = 0.07,
        offset = 0.00000036,
        colnames_position = "top",
        colnames_angle = 90,
        colnames_offset_y = y_offset,
        hjust = 0.5,
        font.size = font_size,
        legend_title = "Plasmid Replicons") +
  scale_fill_viridis_d(aesthetics = "fill", direction = -1) +
    labs(fill = "Presence/Absence") +
  hexpand(.2) +
  vexpand(vertical_expansion)
p11

p11 + theme(legend.key.size = unit(1, "cm")) +
  theme_light(base_size = 10) +
  xlim(NA, 0.0000014) +
  ylim(-0.000001, NA)

ggplotly(p3)
```


```{r}
# save the phages
plotting_dir <- file.path(outdir, Sys.Date())
plot_name <- file.path(plotting_dir, "malawi_first_tree.pdf")

if (!dir.exists(plotting_dir)) {dir.create(plotting_dir, recursive = TRUE)}

pdf(plot_name,
    width = 10,
    height = 20)

p11 + theme(legend.key.size = unit(1, "cm")) +
  theme_light(base_size = 10) +
  xlim(NA, 0.0000014) +
  ylim(-0.000001, NA)


dev.off() 


```
