---
output: html_document
editor_options: 
  chunk_output_type: console
---

# load packages
```{r}
library(tidyverse)
library(kableExtra)
library(stringr)
library(janitor)

outdir <- "outputs/plots"

```

# read and style amr and plasmid data
```{r}

plasmid <-
  read_csv("docs/plasmid_data.csv") %>%
  rename("isolate_id" = "isolate id")

plasmid

reads <-
  read_csv("docs/read_accessions.txt") %>%
  clean_names() %>% rename("isolate_id" = "sample")
 
reads

sequencing_overview <-
  read_csv("docs/sequencing_overview.csv") %>%
  clean_names()

sequencing_overview

long_reads <-
  base::merge(sequencing_overview, reads, by = "isolate_id") %>%
  rename("long_read_accession_number" = "accession") %>%
    select(c(isolate_id, year_of_isolation, source, short_read_accession_number,
             long_read_accession_number, assembly_accession)) %>%
             rename("assembly_accession_number" = "assembly_accession")
  
long_reads
# remove underscore from column names
column_names <- long_reads %>%
  names() %>%
  str_replace_all("_", " ") %>%
  str_to_title()

colnames(long_reads) <- column_names


table_1_data <-
  long_reads %>%
  rename("isolate_id" = "Isolate Id")

table_1_data

read_qc <-
  read_table("/Users/zuza/repos/H58/outputs/nanoplot/nanostats/nanostats.tsv") %>%
    tibble() %>%
  clean_names()

#data for table 2
assembly_stats <-
  sequencing_overview %>%
  select(isolate_id,
           lineage,
           length_of_assembly,
           gc_percent,
           number_of_contigs,
           n50,
           gc_percent,
         busco_completeness_score_percent)

assembly_stats

table_data <-
  merge(read_qc, assembly_stats) %>%
    select(isolate_id,
           number_of_reads,
           mean_read_length,
           mean_read_quality,
           number_of_bases,
           mean_length_n50,
           mean_depth,
           length_of_assembly,
           gc_percent,
           number_of_contigs,
           n50,
           gc_percent,
           busco_completeness_score_percent,
           lineage)

# remove underscore from column names
column_names <-
  table_data %>%
  names() %>%
  str_replace_all("_", " ") %>%
  str_to_title()

# assign the new column names to the dataframe
colnames(table_data) <- column_names

table_2_data_1 <-
  merge(table_data %>%
      rename("isolate_id" = "Isolate Id"),
      plasmid, by = "isolate_id") %>%
      rename("Isolate Id" = "isolate_id")




```


# draw Table 1

```{r}
options(knitr.kable.NA = '')
table_1 <- table_1_data %>%
  rename("Isolate Id" = "isolate_id") %>%
  knitr::kable(booktabs = TRUE,
               longtable = TRUE,
               linesep = "",
               align = "l",
               escape = FALSE) %>%
  kable_styling(position = "center",
                font_size = 15) %>%
  column_spec(1,
              bold = TRUE)


table_1
```

# save to plot plasmid and amr table
```{r}
plotting_dir <- file.path(outdir, Sys.Date())
plot_name <- file.path(plotting_dir, "Table1.pdf")

if (!dir.exists(plotting_dir)) {dir.create(plotting_dir, recursive = TRUE)}

save_kable(table_1, file = plot_name,
           bs_theme = "simplex",
           density = 10000)


```

# draw Table 2

```{r}
options(knitr.kable.NA = '')
#table_2 <-
  table_2_data_1 %>%
  knitr::kable(booktabs = TRUE,
               longtable = TRUE,
               linesep = "",
               align = "l",
               escape = FALSE,
               "html") %>%
  add_header_above(header =
                     c(" "= 1, "Oxford Nanopore Read Stats" = 6,"Assembly Stats"= 5, " " = 1, "Plasmid Replicons" = 3, "AMR Determinants" = 6),
                   font_size = 17) %>%
  kable_styling(position = "center",
                font_size = 14) %>%
  column_spec(1,
              bold = TRUE)


table_2
```

# Table 2A
this table has the read stats and assembly stats
```{r}

#table_2 <-
  table_data %>%
  select(!c(Lineage)) %>%
  knitr::kable(booktabs = TRUE,
               longtable = TRUE,
               linesep = "",
               align = "l",
               escape = FALSE,
               "html") %>%
  add_header_above(header =
                     c(" "= 1, "Oxford Nanopore Read Stats" = 6,"Assembly Stats"= 5),
                   font_size = 17) %>%
  kable_styling(position = "center",
                font_size = 14) %>%
  column_spec(1,
              bold = TRUE)


```


# Table 2B

```{r}

#table2A <-
  plasmid %>%
  mutate(Lineage =
           case_when(plasmid$isolate_id %in%
                       table_data$`Isolate Id` ~
                       table_data$Lineage)) %>%
  select(c(isolate_id,
           Lineage,
           IncFIAHI1,
           IncHI1A,
           IncHI1BR27,
           gyrA_S83F,
           `blaTEM-1D`,
           sul2,
           dfrA14,
           tetB,
           `aph(3'')-Ib`)) %>%
  knitr::kable(booktabs = TRUE,
               longtable = TRUE,
               linesep = "",
               align = "l",
               escape = FALSE,
               "html") %>%
  add_header_above(header =
                     c(" "= 2,
                       "Plasmid Replicons" = 3,
                       "AMR Determinants" = 6),
                   font_size = 17) %>%
  kable_styling(position = "center",
                font_size = 14) %>%
  column_spec(1,
              bold = TRUE)

```



# QC plots

```{r}
library(scales)
reads_col <- "#01665e"
assemblies_col = "#440154"
y_axis_size = 10
y_lab_size = 11

plot_data <-table_data %>%
  clean_names()

read_length <-
  plot_data %>%
  ggplot() +
  geom_col(aes(isolate_id, mean_read_length),
           fill = reads_col) +
    ylab("Mean Read Length") +
      theme(axis.text.y = element_text(size = y_axis_size,
                                 face = "bold"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(face = "bold",
                                  size = y_lab_size))

mean_read_quality <-
  plot_data %>%
  ggplot() +
  geom_col(aes(isolate_id, mean_read_quality),
           fill = reads_col) +
    ylab("Mean Read Phred Quality Score") +
    ylim(0,15) +
    scale_y_continuous(n.breaks = 10) +
      theme(axis.text.y = element_text(size = y_axis_size,
                                 face = "bold"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(face = "bold",
                                  size = y_lab_size))

number_of_bases <-
  plot_data %>%
  ggplot() +
  geom_col(aes(isolate_id, number_of_bases),
           fill = reads_col) +
    ylab("Total Number of Bases") +
    scale_y_continuous(labels = label_bytes(),
                       limits = c(0, 400000000),
                       n.breaks = 10) +
      theme(axis.text.y = element_text(size = y_axis_size,
                                 face = "bold"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(face = "bold",
                                  size = y_lab_size))

num_reads <-
  plot_data %>%
  ggplot() +
  geom_col(aes(isolate_id, number_of_reads),
           fill = reads_col) +
    ylab("Number of Reads") +
    ylim(0,40000) +
      theme(axis.text.y = element_text(size = y_axis_size,
                                 face = "bold"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(face = "bold",
                                  size = y_lab_size))

mean_length_n50 <-
  plot_data %>%
  ggplot() +
  geom_col(aes(isolate_id, mean_length_n50),
           fill = reads_col) +
    ylab("Mean Read N50") +
    ylim(0,15000) +
      theme(axis.text.y = element_text(size = y_axis_size,
                                 face = "bold"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(face = "bold",
                                  size = y_lab_size))

mean_depth <-
  plot_data %>%
  ggplot() +
  geom_col(aes(isolate_id, mean_depth),
           fill = reads_col) +
    ylab("Mean Read Coverage Depth") +
    ylim(0,80) +
      theme(axis.text.y = element_text(size = y_axis_size,
                                 face = "bold"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(face = "bold",
                                  size = y_lab_size))

length_of_assembly <-
  plot_data %>%
  ggplot() +
  geom_col(aes(isolate_id, length_of_assembly), fill = assemblies_col) +
    ylab("Total Length of Assembly") +
    scale_y_continuous(labels = label_bytes(),
                       n.breaks = 5) +
      theme(axis.text.y = element_text(size = y_axis_size,
                                 face = "bold"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(face = "bold",
                                  size = y_lab_size))

gc_percent <-
  plot_data %>%
  ggplot() +
  geom_col(aes(isolate_id, gc_percent),
           fill = assemblies_col) +
    ylab("GC%") +
    scale_y_continuous(n.breaks = 15, limits = c(0, 60)) +
      theme(axis.text.y = element_text(size = y_axis_size,
                                 face = "bold"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(face = "bold",
                                  size = y_lab_size))

number_of_contigs <-
  plot_data %>%
  ggplot() +
  geom_col(aes(isolate_id, as.factor(number_of_contigs)), fill = assemblies_col) +
    ylab("Number of Contigs") +
      theme(axis.text.y = element_text(size = y_axis_size,
                                 face = "bold"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(face = "bold",
                                  size = y_lab_size))

n50 <-
  plot_data %>%
  ggplot() +
  geom_col(aes(isolate_id, n50),
           fill = assemblies_col) +
    ylab("N50") +
    scale_y_continuous(labels = label_bytes(),
                       n.breaks = 5) +
      theme(axis.text.y = element_text(size = y_axis_size,
                                 face = "bold"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(face = "bold",
                                  size = y_lab_size))

busco_completeness_score_percent <-
  plot_data %>%
  ggplot() +
  geom_col(aes(isolate_id,
               as.integer( busco_completeness_score_percent)), fill = assemblies_col) +
    ylab("BUSCO Completeness Score (%)") +
    scale_y_continuous(n.breaks = 10) +
      theme(axis.text.y = element_text(size = y_axis_size,
                                 face = "bold"),
        axis.title.x = element_blank(),
        axis.title.y = element_text(face = "bold",
                                  size = y_lab_size))


read_length + mean_length_n50 +
  mean_read_quality + number_of_bases +
  num_reads + mean_length_n50 + mean_depth +
  length_of_assembly + n50 +
  gc_percent + number_of_contigs +
  busco_completeness_score_percent +
  plot_annotation(title =
                    "ONT read and hybrid assembly statistics",
                  theme =
                    theme(plot.title =
                            element_text(size = 15,
                                         face = "bold",
                                         hjust = 0.5)))

```





#save table 2
```{r}
plotting_dir <- file.path(outdir, Sys.Date())
plot_name <- file.path(plotting_dir, "Table2.pdf")

if (!dir.exists(plotting_dir)) {dir.create(plotting_dir, recursive = TRUE)}

save_kable(table_2, file = plot_name,
           bs_theme = "simplex")

```


# Save table 1
```{r}
plotting_dir <- file.path(outdir, Sys.Date())
plot_name <- file.path(plotting_dir, "Table1.pdf")

if (!dir.exists(plotting_dir)) {dir.create(plotting_dir, recursive = TRUE)}

save_kable(table_1, file = plot_name,
           bs_theme = "simplex")
```












```{r}

sequencing_overview <-
  read_csv("docs/sequencing_overview.csv") %>%
  clean_names() %>%
  filter(isolate_id != "A53789")

# calculate summary values
mean_length <- mean(sequencing_overview$length_of_assembly)

mean_gc <- mean(sequencing_overview$gc_percent)

table <- sequencing_overview %>% select(isolate_id,
                                        year_of_isolation,
                                        source,
                               long_read_accession,
                               short_read_accession_number,
                               lineage,
                               length_of_assembly,
                               n50,
                               number_of_contigs,
                               gc_percent,
                               presence_of_plasmid)

# remove underscore from column names
column_names <- table %>%
  names() %>%
  str_replace_all("_", " ") %>%
  str_to_title()


# draw the table
#y <-
  table %>%
  knitr::kable(booktabs = TRUE,
               longtable = TRUE,
               linesep = "",
               align = "l",
               escape = FALSE,
    col.names = linebreak(column_names, align = "l")) %>%
  kable_styling(position = "left",
                full_width = FALSE,
                latex_options = c("striped", "repeat_header"),
                stripe_color = "gray")
```


# save isolate metadata to table
```{r}
plotting_dir <- file.path(outdir, Sys.Date())
plot_name <- file.path(plotting_dir, "Table1.png")

if (!dir.exists(plotting_dir)) {dir.create(plotting_dir, recursive = TRUE)}

save_kable(y, file = plot_name,
           bs_theme = "simplex")

```
